keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -395,15 +396,36 @@ class SqlArtifactRepository(
         .execute()
     }
     sqlRetry.withRetry(WRITE) {
+      // update old "CURRENT" to "PREVIOUS, set promotionReference for use in summary data
       jooq
         .update(ENVIRONMENT_ARTIFACT_VERSIONS)
         .set(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS, PREVIOUS.name)
+        .set(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_REFERENCE, version)
         .where(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(environmentUid))
         .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID.eq(artifact.uid))
         .and(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS.eq(CURRENT.name))
         .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION.ne(version))
         .execute()
     }
+    sqlRetry.withRetry(WRITE) {
+      // update any past artifacts that were "APPROVED" to be "SKIPPED
+      // because the new version takes precedence
+      val approvedButOld = jooq.select(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION)
+        .from(ENVIRONMENT_ARTIFACT_VERSIONS)
+        .where(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS.eq(APPROVED.name))
+        .fetch(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION)
+        .filter { artifact.versioningStrategy.comparator.compare(it, version) > 0 } // get the lower versions
```

Comment:
This is very confusing to read because the contract of any `Comparator` is to return `> 0` when the item on the left is higher than the one on the right, but because we've hard-coded the version strategy comparators to be descending by default, you have to use that in the backwards sense here. How about adding an `ascendingComparator` to `versionStrategy` which returns `comparator.inverse()` so that the original semantics is maintained? Even better, if you could add helper methods like `versioningStrategy.isHigher(ver1, ver2)` and `versioningStrategy.isLower(ver1, ver2)`, that would make it much more readable IMHO.

Reply:
I'm going to add a helper method to just the sql artifact repository to make it easier to read, but not have to do a big change. It's not super straightforward to add helper methods onto versioning strategy.